### PR TITLE
(Streamline delivery) Add delivery table and edit via PUT and PATCH

### PR DESCRIPTION
**Pull request by islean**
## Description

We wish to stop marking analyses as delivered by writing comments. This PR adds a table for tracking 
- Who delivered the analysis
- When it was delivered

Closes https://github.com/Clinical-Genomics/streamline-delivery/issues/67.

### Added

- Delivery table

### Changed

- You can set analyses as delivered/undelivered in /analyses and /analyses/<id>

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b add-delivery-table -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by islean** [comment]
> Cool - there is a delivery table in statusdb: https://cg-statusdb.scilifelab.se/admin/delivery/. What is the plan for that one (in relation to this one)? Can it be dropped? Would be nice to keep it DRY.

A bit uncertain how that table is used, I will look into it. If it is not used, we can probably drop it? Given that the table in this PR is intended to track analyses in Trailblazer, it feels natural to keep it here.

**Comment by islean** [comment]
Deployment to stage successful:

```
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/isak.ohlsson
done.
Log deploy... done.
trailblazer, version 21.2.2
θ62° [isak.ohlsson@hasta:~] [S_base] 2m30s $ 
```

Table created:

image

Correct constraints and foreign keys:

image

image

**Comment by islean** [comment]
Should be delivered_date instead of delivered_at

File: trailblazer/store/models.py [hunk 1]
```diff
@@ -134,6 +143,21 @@ def to_dict(self) -> dict:
...
+    __tablename__ = "delivery"
+    __table_args__ = (UniqueConstraint("analysis_id"),)
+
+    id = Column(types.Uuid, primary_key=True)
+    analysis_id = Column(ForeignKey(Analysis.id, ondelete="CASCADE"), nullable=False)
+    delivered_by = Column(ForeignKey(User.id), nullable=False)
+    delivered_date = Column(types.DateTime, nullable=False)
+
+    analysis = orm.relationship("Analysis", back_populates="delivery")
+    user = orm.relationship("User")
```
Comment (seallard) [inline trailblazer/store/models.py:158]: Can/should we mark all of these as non-nullable?

**Review by seallard — APPROVED** [review]
Nice 👍  Maybe we can mark some of them as non nullable

**Comment by islean** [comment]
Fixed delivered_at -> delivered_date and type is now date.

File: trailblazer/store/models.py (hunk 1 above)
Comment (seallard) [reply trailblazer/store/models.py:158]: Missed that all fields that can be non nulleable already are marked as such

**Comment by islean** [comment]
Tested adding and deleting the entries directly in the database. The deletion did not cascade to Analysis nor User. Foreign keys work. MYSQL does not support UUID, so the database itself can not check if the id is globally unique, but my suspicion is that everything added via our codebase will have a universally unique id.

File: trailblazer/dto/update_analyses.py [hunk 2]
```diff
@@ -5,9 +5,11 @@
 
 class AnalysisUpdate(BaseModel):
     id: int
-    status: TrailblazerStatus | None = None
+
     comment: str | None = None
+    delivered: bool | None = None
```
Comment (ChrOertlin) [inline trailblazer/dto/update_analyses.py:10]:
```suggestion
    is_delivered: bool | None = None
```

Comment (ChrOertlin) [reply trailblazer/dto/update_analyses.py:10]: bool pre-fix `is_`

File: trailblazer/store/crud/update.py [hunk 3]
```diff
@@ -250,6 +270,10 @@ def update_analysis(
             LOG.info(f"Adding comment {comment} to analysis {analysis.id}")
             analysis.comment = comment
 
+        if delivered is not None:
```
Comment (ChrOertlin) [inline trailblazer/store/crud/update.py:273]:
```suggestion
        if is_delivered is not None:
```

File: trailblazer/store/crud/update.py [hunk 4]
```diff
@@ -250,6 +270,10 @@ def update_analysis(
...
+            LOG.info(f"Setting analysis delivered status to {delivered}")
```
Comment (ChrOertlin) [inline trailblazer/store/crud/update.py:274]:
```suggestion
            LOG.info(f"Setting analysis delivered status to {is_delivered}")
```

File: trailblazer/store/models.py [hunk 5]
```diff
@@ -108,11 +110,22 @@ def upload_jobs(self) -> list["Job"]:
...
+    @property
+    def delivered_date(self) -> datetime.datetime | None:
+        return self.delivery.delivered_date if self.delivery else None
+
     def to_dict(self) -> dict:
         """Return a dictionary representation of the object."""
         return {
             "id": self.id,
             "case_id": self.case_id,
+            "delivered": bool(self.delivery),
```
Comment (ChrOertlin) [inline trailblazer/store/models.py:126]:
```suggestion
            "is_delivered": bool(self.delivery),
```

File: trailblazer/dto/analyses_response.py [hunk 6]
```diff
@@ -18,6 +18,7 @@ class Analysis(BaseModel):
     comment: str | None = None
     completed_at: datetime | None = None
     config_path: str | None = None
+    delivered: bool | None = None
```
Comment (ChrOertlin) [inline trailblazer/dto/analyses_response.py:21]: Maybe we can also just initialise as False, then we can skip none checks? I mean ideally it is either true or false

Comment (ChrOertlin) [inline trailblazer/dto/analyses_response.py:21]:
```suggestion
    is_delivered: bool | None = None
```

**Comment by islean** [comment]
Tested sending requests to the /analyses PATCH and to the /analyses/<analysis_id> PUT and both were able to add an entry and drop the entry.

File: trailblazer/dto/analyses_response.py (hunk 6 above)
Comment (islean) [reply trailblazer/dto/analyses_response.py:21]: False currently means that the existing entry gets dropped, so I don't think that is an option currently 😞

File: trailblazer/dto/update_analyses.py (hunk 2 above)
Comment (islean) [reply trailblazer/dto/update_analyses.py:10]: Good suggestion, I'll change it.

Comment (islean) [reply trailblazer/dto/update_analyses.py]: Changed.

**Comment by islean** [comment]
Tested with single analysis, multiple analyses, true/false, combination of true/false for PATCH, and True on already delivered and False on not delivered. Worked as expected, i.e.

is_delivered: True & analysis already delivered -> no action
is_delivered: True & analysis not already delivered -> adds a Deliver entry
is_delivered: False & analysis already delivered -> drop entry
is_delivered: False & analysis not already delivered -> no action

Bulk works exactly the same but loops through the analyses.

File: trailblazer/dto/update_analyses.py [hunk 7]
```diff
@@ -5,9 +5,10 @@
 
 class AnalysisUpdate(BaseModel):
     id: int
-    status: TrailblazerStatus | None = None
     comment: str | None = None
+    is_delivered: bool | None = None
```
Comment (seallard) [inline trailblazer/dto/update_analyses.py:9]: Nice ⭐

File: trailblazer/server/api.py [hunk 8]
```diff
@@ -111,7 +101,10 @@ def patch_analyses(analysis_service: AnalysisService = Provide[Container.analysi
     """Update data (such as status, visibility, comments etc.) for multiple analyses at once."""
     try:
         request_data = UpdateAnalyses.model_validate(request.json)
-        response: UpdateAnalysesResponse = analysis_service.update_analyses(request_data)
+        user: User = g.get("current_user")
+        response: UpdateAnalysesResponse = analysis_service.update_analyses(
+            data=request_data, user=user
```
Comment (seallard) [inline trailblazer/server/api.py:106]: I'd consider passing down the user id, can't really give a good motivation for it though so take it with a grain of salt 😅 

I guess I don't like passing around database records in the api layer

File: trailblazer/store/crud/update.py [hunk 9]
```diff
@@ -233,12 +233,32 @@ def update_analysis_comment(self, case_id: str, comment: str) -> None:
         session.commit()
         LOG.info(f"Adding comment {comment} to analysis {analysis.case_id}")
 
+    @staticmethod
+    def update_analysis_delivery(
+        analysis: Analysis, is_delivered: bool, user: User | None = None
+    ) -> None:
+        session: Session = get_session()
+        if is_delivered and not analysis.delivery:
```
Comment (seallard) [inline trailblazer/store/crud/update.py:241]: Why do we need the `not analysis.delivery` check?

File: trailblazer/store/models.py [hunk 10]
```diff
@@ -134,6 +147,21 @@ def to_dict(self) -> dict:
...
+class Delivery(Model):
+    """Tracks when and by whom an analysis was delivered."""
+
+    __tablename__ = "delivery"
+    __table_args__ = (UniqueConstraint("analysis_id"),)
+
+    id = Column(types.Uuid, primary_key=True, default=uuid.uuid4)
+    analysis_id = Column(ForeignKey(Analysis.id, ondelete="CASCADE"), nullable=False)
+    delivered_by = Column(ForeignKey(User.id), nullable=False)
+    delivered_date = Column(types.DateTime, nullable=False)
```
Comment (seallard) [inline trailblazer/store/models.py:159]: `delivery_date`?

**Review by seallard — APPROVED** [review]
Nice 🥳 Looks great!

File: trailblazer/store/crud/update.py (hunk 9 above)
Comment (islean) [reply trailblazer/store/crud/update.py:241]: Because it crashes if we try add another entry to the database - they should only be delivered once right?

Comment (islean) [reply trailblazer/store/crud/update.py:241]: Double checked with @beatrizsavinhas and seems as if this is good for now, but we can think of whether there is a benefit to being able to track multiple deliveries.

**Comment by islean** [comment]
Deployed to Hasta and cg-services-prod

**Pull request merged**